File: Data Structures & Algorithms/maximum-product-subarray/submission-1.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        int n = nums.size();
        int ans = nums[0];

        int mn = nums[0], mx = nums[0];

        for(int i=1; i<n; i++){
            int v = nums[i];

            if(v<0) swap(mn, mx);

            mx = max( v, v*mx);
            mn = min( v, v*mn);

            ans = max( ans, mx);
        }

        return ans;
    }
};

File: Data Structures & Algorithms/maximum-subarray/submission-1.java
class Solution {
    public int maxSubArray(int[] nums) {
        int mx = Integer.MIN_VALUE, cur = 0;

        for(int i:nums){
            cur = Math.max(cur+i, i);
            mx = Math.max(mx, cur);
        }
        return mx;
    }
}
